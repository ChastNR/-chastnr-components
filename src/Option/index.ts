export { Options } from './Options';
export { OptionsContainer } from './Container';
export { SelectOption } from './Option';
