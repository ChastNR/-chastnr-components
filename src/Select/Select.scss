.ctrl__cntr {
  &.slt {
    cursor: pointer;
  }

  .slt__value {
    line-height: 0;
    outline: 0;
    padding: 22px 0 9px;
    position: relative;
    text-overflow: ellipsis;
    transition: opacity 0.25s ease;
    white-space: nowrap;
    width: 100%;
    will-change: opacity;
    word-wrap: break-word;
  }

  &:focus-within {
    .arrow {
      transform: rotate(180deg);
    }

    .slt__opts {
      height: auto;
      opacity: 1 !important;
      z-index: 100;
    }
  }
}
