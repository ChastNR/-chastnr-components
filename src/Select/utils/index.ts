export { createSelectedItems } from './createSelectedItems';
export { createSelectedOptions } from './createSelectedOptions';
export { createSelectedValue } from './createSelectedValue';
