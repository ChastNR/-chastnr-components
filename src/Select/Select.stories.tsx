/* eslint-disable import/no-extraneous-dependencies */
import { useState } from 'react';
import { ComponentMeta } from '@storybook/react';
import { IOption } from '../types';
import { Select } from './Select';

const componentMeta: ComponentMeta<typeof Select> = {
  title: 'Select',
  component: Select,
  parameters: {
    layout: 'fullscreen',
  },
};

const testOptions: IOption[] = [{ title: 'Test 1', value: 'Test1' }];

export const DefaultSelect = () => {
  const [value, setValue] = useState<IOption>();

  return (
    <Select
      label="Выберите файлы или перетяните их сюда"
      onSelect={setValue}
      options={testOptions}
      selected={value}
    />
  );
};

export default componentMeta;
