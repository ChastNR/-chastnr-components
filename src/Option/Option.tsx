import { useCallback, useState } from "react";

import Checkbox from "../Checkbox";
import { Option } from "../types";
import { mouseDownPrevent } from "../utils";

interface OptionProps {
  option: Option;
  checked?: boolean;
  isMulti?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const SelectOption: React.FC<OptionProps> = ({ checked = false, option, isMulti, onClick }) => {
  const [isChecked, setChecked] = useState(checked);

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      setChecked((s) => !s);

      if (onClick) {
        onClick(e);
      }
    },
    [onClick]
  );

  return (
    <button className="slt__opt" onClick={handleClick} onMouseDown={mouseDownPrevent} type="button">
      <div className="slt__opt__txt">{option.title}</div>
      {isMulti && <Checkbox checked={isChecked} />}
    </button>
  );
};

export default SelectOption;
