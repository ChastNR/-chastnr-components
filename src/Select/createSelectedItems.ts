import { Option } from "../Option";

const createSelectedItems = (options: Option[], selected: Option[], separator = ", "): string => {
  const selectedTitles = options.reduce((acc: React.ReactNode[], o) => {
    const option = selected.find((s) => o.value === s.value);

    if (option) {
      acc.push(option.title);
    }

    return acc;
  }, []);

  return selectedTitles.join(separator);
};

export default createSelectedItems;
