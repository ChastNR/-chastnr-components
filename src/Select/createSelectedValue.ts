import { Option } from "../Option";

const createSelectedValue = (selected?: Option | Option[], isMulti?: boolean): Option[] => {
  if (!selected) {
    return [];
  }

  if (isMulti) {
    return selected as Option[];
  }

  return [selected as Option];
};

export default createSelectedValue;
