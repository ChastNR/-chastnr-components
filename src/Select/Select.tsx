/* eslint-disable jsx-a11y/control-has-associated-label */
import { useCallback, useRef, useState } from "react";

import { Arrow, createError } from "../base";
import Control from "../Control";
import { Option, Options } from "../Option";
import createSelectedItems from "./createSelectedItems";

import "./Select.scss";

interface SelectProps {
  disabled?: boolean;
  errorMessage?: React.ReactNode;
  isMulti?: boolean;
  label: React.ReactNode;
  onSelect: (value: Option | Option[]) => void;
  options: Option[];
  optionsSeparator?: string;
  requiredMessage?: React.ReactNode;
  selected?: Option[];
}

const Select: React.FC<SelectProps> = ({
  disabled,
  errorMessage,
  isMulti = false,
  label,
  onSelect,
  options,
  requiredMessage,
  optionsSeparator = ", ",
  selected = [],
}) => {
  const wrapperRef = useRef<HTMLDivElement>(null);

  const [validate, setValidate] = useState(false);

  const handleBlur = useCallback((e: React.FocusEvent<HTMLDivElement>) => {
    if (e.currentTarget.contains(e.relatedTarget as Node)) {
      return;
    }

    setValidate(true);
  }, []);

  const handleSelect = (option: Option) => () => {
    if (isMulti) {
      const isOptionSelected = selected.some((o) => o.value === option.value);

      const selectedOptions = isOptionSelected
        ? selected.filter((o) => o.value !== option.value)
        : selected.concat(option);

      onSelect(selectedOptions);
    } else {
      wrapperRef.current?.blur();
      onSelect(option);
    }

    setValidate(false);
  };

  const selectedItems = createSelectedItems(options, selected, optionsSeparator);

  const hasValue = selected.length > 0;

  const error = createError({
    validate,
    requiredMessage,
    value: hasValue,
    errorMessage,
  });

  return (
    <Control
      className="slt"
      content={<Arrow />}
      disabled={disabled}
      error={error}
      hasValue={hasValue}
      label={label}
      onBlur={handleBlur}
      ref={wrapperRef}
      required={!!requiredMessage}
    >
      <div className="slt__value">{selectedItems}</div>
      <Options isMulti={isMulti} onClick={handleSelect} options={options} />
    </Control>
  );
};

export default Select;
