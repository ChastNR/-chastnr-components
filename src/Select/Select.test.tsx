import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Option } from '../Option';
import Select from './Select';

describe('<Select/>', () => {
  const options: Option[] = [];
  const onSelectMock = jest.fn();

  it('should match snapshot', () => {
    const select = shallow(
      <Select label="Test Select" onSelect={onSelectMock} options={options} />
    );
    expect(toJson(select)).toMatchSnapshot();
  });

  it('should render options', () => {
    const select = mount(<Select label="Test Select" onSelect={onSelectMock} options={options} />);
    const optionsComponent = select.find('Options');
    expect(optionsComponent).toBeTruthy();
  });

  it('props should be equal', () => {
    const label = 'Test Select';
    const disabled = true;
    const name = 'testName';
    const errorMessage = 'Error text';
    const requiredMessage = 'Required text';
    const isMulti = true;
    const selected = [options[0]];

    const select = mount(
      <Select
        disabled={disabled}
        errorMessage={errorMessage}
        isMulti={isMulti}
        label={label}
        name={name}
        onSelect={onSelectMock}
        options={options}
        requiredMessage={requiredMessage}
        selected={selected}
      />
    );

    expect(select.props().label).toEqual(label);
    expect(select.props().disabled).toEqual(disabled);
    expect(select.props().errorMessage).toEqual(errorMessage);
    expect(select.props().isMulti).toEqual(isMulti);
    expect(select.props().requiredMessage).toEqual(requiredMessage);
    expect(select.props().selected.length).toEqual(1);
  });
});
