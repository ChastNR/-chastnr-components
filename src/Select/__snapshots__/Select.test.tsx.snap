// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> should match snapshot 1`] = `
<ForwardRef
  className="slt"
  content={<Memo />}
  error={null}
  hasValue={false}
  label="Test Select"
  onBlur={[Function]}
  required={false}
>
  <div
    className="slt__value"
  />
  <Options
    isMulti={false}
    onClick={[Function]}
    options={Array []}
  />
</ForwardRef>
`;
